src/libraries/Common/src/System/Reflection/Metadata/TypeNameParserHelpers.cs

```diff
@@ -0,0 +1,747 @@
+                {
+                    if (backtickIndex == 0)
+                    {
+                        return -1; // illegal name, example "`1"
+                    }
+                    else if (fullTypeName[backtickIndex - 1] == EscapeCharacter)
+                    {
+                        return 0; // legal name, but not a generic type definition. Example: "Escaped\\`1"
+                    }
+                    else if (TryParse(fullTypeName.Slice(backtickIndex + 1), out int value))
+                    {
+                        // From C# 2.0 language spec: 8.16.3 Multiple type parameters Generic type declarations can have any number of type parameters.
+                        if (value > MaxLength)
+                        {
+                            // But.. it's impossible to create a type with more than Array.MaxLength.
+                            // OOM is also not welcomed in the parser!
+                            return -1;
+                        }
+
+                        // the value can still be negative, but it's fine as the caller should treat that as an error
+                        return value;
+                    }
+
+                    // most likely the value was too large to be parsed as an int
+                    return -1;
+                }
+                else if (!IsAsciiDigit(fullTypeName[backtickIndex]) && fullTypeName[backtickIndex] != '-')
+                {
+                    break;
+                }
+            }
+
+            return 0;
+        }
+
+        internal static string GetGenericTypeFullName(ReadOnlySpan<char> fullTypeName, TypeName[]? genericArgs)
+        {
+            if (genericArgs is null)
+            {
+                return fullTypeName.ToString();
+            }
+
+            ValueStringBuilder result = new(stackalloc char[128]);
+            result.Append(fullTypeName);
+
+            result.Append('[');
+            foreach (TypeName genericArg in genericArgs)
+            {
+                result.Append('[');
+                result.Append(genericArg.AssemblyQualifiedName);
+                result.Append(']');
+                result.Append(',');
+            }
+            result[result.Length - 1] = ']'; // replace ',' with ']'
+
+            return result.ToString();
+        }
+
+        /// <returns>Positive length or negative value for invalid name</returns>
+        internal static int GetFullTypeNameLength(ReadOnlySpan<char> input, out bool isNestedType)
+        {
+            isNestedType = false;
+
+            // NET 6+ guarantees that MemoryExtensions.IndexOfAny has worst-case complexity
+            // O(m * i) if a match is found, or O(m * n) if a match is not found, where:
+            //   i := index of match position
+            //   m := number of needles
+            //   n := length of search space (haystack)
+            //
+            // Downlevel versions of .NET do not make this guarantee, instead having a
```

Comment:
This is specifically about MemoryExtensions.IndexOfAny in the System.Memory.dll nuget package, right?

Reply:
Correct.

https://github.com/dotnet/maintenance-packages/blob/d4261b1112437567ca4841e0c251a0ec914e7213/src/System.Memory/src/System/SpanHelpers.T.cs#L298-L309